guava/src/com/google/common/collect/ImmutableSet.java

```diff
@@ -193,23 +193,41 @@
    * @throws NullPointerException if any of {@code elements} is null
    * @since 7.0 (source-compatible since 2.0)
    */
+  @SuppressWarnings("unchecked") // all supported methods are covariant
   public static <E> ImmutableSet<E> copyOf(Collection<? extends E> elements) {
     /*
      * TODO(lowasser): consider checking for ImmutableAsList here
      * TODO(lowasser): consider checking for Multiset here
      */
     // Don't refer to ImmutableSortedSet by name so it won't pull in all that code
     if (elements instanceof ImmutableSet && !(elements instanceof SortedSet)) {
-      @SuppressWarnings("unchecked") // all supported methods are covariant
       ImmutableSet<E> set = (ImmutableSet<E>) elements;
       if (!set.isPartialView()) {
         return set;
+      } else {
+        // We use the iterator to avoid making a copy of the set, since inserting into the set would
+        // make a copy of the collection anyways under this condition.
+
+        // we know that sets are all unique, so we can supply the set size to avoid resizing.
+        return construct(elements.size(), elements.size(), (Iterator<Object>) elements.iterator());
       }
     } else if (elements instanceof EnumSet) {
       return copyOfEnumSet((EnumSet) elements);
+    } else if (elements instanceof ImmutableCollection) {
+      // The collection is immutable, so avoid making a copy of the collection right now since
+      // inserting into the set would make a copy of the elements anyways under this condition
+      return construct(elements.size(), (Iterator<Object>) elements.iterator());
+    } else if (elements instanceof Set) {
+      // we know that sets are all unique, so we can supply the set size to avoid resizing.
+
+      // In case of mutation, make a copy first for thread safety.
+      Object[] array = elements.toArray();
+      return construct(array.length, array.length, Iterators.forArray(array));
     }
+
+    // In case of mutation, make a copy first for thread safety.
     Object[] array = elements.toArray();
```

Comment:
Previously, even when we have an ImmutableCollection, we would create a copy of the collection as an array, and then insert them into the Builder, which would create an another array, which unnecessarily creates an array.
This made for a regression in the copyOf(ImmutableSet) benchmark.
Creating an iterator for the ImmutableCollection is much more efficient than copying the array.

I changed everything else to use iterators so that we can have one consistent api for `construct()`. And profiling didn't show that big a difference.
Otherwise, we need two `construct()` implementations where one takes in an `Iterable` and another takes in an `Array`.